component/registry.go

```diff
@@ -0,0 +1,240 @@
+package component
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/regexp"
+	"github.com/hashicorp/hcl/v2"
+)
+
+// The parsedName of a component is the parts of its name ("remote.http") split
+// by the "." delimiter.
+type parsedName []string
+
+// String re-joins the parsed name by the "." delimiter.
+func (pn parsedName) String() string { return strings.Join(pn, ".") }
+
+var (
+	// Globally registered components
+	registered = map[string]Registration{}
+	// Parsed names for components
+	parsedNames = map[string]parsedName{}
+)
+
+// Options are provided to a component when it is being constructed. Options
+// are static for the lifetime of a component.
+type Options struct {
+	// ID of the component. Guaranteed to be globally unique across all running
+	// components.
+	ID string
+
+	// Logger the component may use for logging. The component ID will always be
+	// set as a field.
+	Logger log.Logger
+
+	// A path to a directory with this component may use for storage. The path is
+	// guaranteed to be unique across all running components.
+	//
+	// The directory may not exist when the component is created; components
+	// should create the directory if needed.
+	DataPath string
+
+	// OnStateChange may be invoked at any time by a component whose Export value
+	// changes. The Flow controller then will queue re-processing components
+	// which depend on the changed component.
+	//
+	// OnStateChange will panic if e does not match the Exports type registered
+	// by the component; a component must use the same Exports type for its
+	// lifetime.
+	OnStateChange func(e Exports)
+}
+
+// Registration describes a single component.
+type Registration struct {
+	// Name of the component. Must be a list of period-delimited valid
+	// identifiers, such as "remote.s3". Components sharing a prefix must have
+	// the same number of identifiers; it is valid to register "remote.s3" and
+	// "remote.http" but not "remote".
+	//
+	// Components may not have more than 2 identifiers.
+	//
+	// Each identifier must start with a valid ASCII letter, and be followed by
+	// any number of underscores or alphanumeric ASCII characters.
+	Name string
+
+	// A singleton component only supports one instance of itself across the
+	// whole process. Normally, multiple components of the same type may be
+	// created.
+	//
+	// The fully-qualified name of a component is the combination of HCL block
+	// name and all of its labels. Fully-qualified names must be unique across
+	// the process. Components which are *NOT* singletons automatically support
+	// user-supplied identifiers:
```

Comment:
I wonder if we should leave this as-is and refactor as you suggested once we know whether we'll want more labels. When I was thinking of bringing in OTel components, it seemed like having a third label would've been useful (i.e., otel.receiver.kafka instead of otel.receiver_kafka)